posts/2024-09-05-Rust-1.81.0.md

```diff
@@ -0,0 +1,160 @@
+
+If you'd like to help us out by testing future releases, you might consider updating locally to use the beta channel (`rustup default beta`) or the nightly channel (`rustup default nightly`). Please [report](https://github.com/rust-lang/rust/issues/new/choose) any bugs you might come across!
+
+## What's in 1.81.0 stable
+
+### `core::error::Error`
+
+1.81 stabilizes the `Error` trait in `core`, allowing usage of the trait in `#![no_std]` libraries.
+
+TODO: What can we actually say about this? It seems like this mostly enables
+things in the ecosystem, but from the langauge/standard library not much has
+actually changed in *this* release?
+
+### New sort implementations
+
+Both the stable and unstable sort implementations in the standard library have
+been updated to new algorithms, improving their runtime performance and
+compilation time. Both of the new sort algorithms try to detect
+incorrect implementations of `Ord` and will panic on such cases if detected.
+
+Users encountering these panics should audit any custom Ord implementations to
+ensure they satisfy the requirements documented in [PartialOrd] and [Ord].
+
+TODO: Ideally we'd have text somewhere specific to the new detection that helps
+explain how to do this audit or otherwise helps users. This otherwise feels
+pretty opaque to me. We might also want to consider whether some kind of
+transition rather than panic! is best (not sure how else users would find out
+about the problem though). See https://github.com/rust-lang/rust/issues/129561.
+
+[PartialOrd]: https://doc.rust-lang.org/nightly/std/cmp/trait.PartialOrd.html
+[Ord]: https://doc.rust-lang.org/nightly/std/cmp/trait.Ord.html
+
+### `#[expect(lint)]`
+
+1.81 stabilizes a new lint level, `expect`, which allows explicitly noting that
+a particular lint *should* occur, and warning if it doesn't.  The intended use
+case for this is temporarily silencing a lint, whether due to lint
+implementation bugs or ongoing refactoring, while wanting to know when the lint
+is no longer required.
+
+For example, if you're moving a code base to comply with a new restriction
+enforced via a Clippy lint like
+[`undocumented_unsafe_blocks`](https://rust-lang.github.io/rust-clippy/stable/index.html#/undocumented_unsafe_blocks),
+you can use `#[expect(clippy::undocumented_unsafe_blocks)]` as you transition,
+ensuring that once all unsafe blocks are documented you can opt into denying
+the lint to enforce it.
+
+### Lint reasons
+
+Changing the lint level is often done for some particular reason. For example,
+if code runs in an environment without floating point support, you could use
+Clippy to lint on such usage with `#![deny(clippy::float_arithmetic)]`.
+However, if a new developer to the project sees this lint fire, they need to
+look for (hopefully) a comment on the deny explaining why it was added. With
+Rust 1.81, they can be informed directly in the compiler message:
```

Comment:
I think it would be good to include an example in Rust showing what this looks like (instead of relying on just the error message output).

````suggestion
Rust 1.81, they can be informed directly in the compiler message by adding a `reason` to the lint attribute:

```rust
#![deny(clippy::float_arithmetic, reason = "no hardware float support")]
```

When this lint is triggered, the compiler will include the reason in its output:
````